### ISSUE TITLE
[css-overflow-3] 'overflow' 2-value syntax is in wrong order

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#2988]
Currently the 'overflow' 2-value syntax is parsed to be horizontal-vertical (to be precise, the first value maps to overflow-x, the second to overflow-y).

When possible, for any new 2-axis properties we're trying to map them to the logical longhands instead, so this is currently wrong.  As well, for horizontal writing modes, the x/y ordering is *opposite* the logical ordering we've settled on, which is block/inline (to match what 4-value syntaxes do when you specify only two values).

As the two-value syntax is new, this should still be fixable - two values should map to overflow in the block and inline axises, in that order.

**Comment by fantasai** [comment]
It's especially important to keep `scroll-snap-align` and `overflow` consistent since they will often be used together.

**Comment by dbaron** [comment]
I filed [Mozilla bug 1481866](https://bugzilla.mozilla.org/show_bug.cgi?id=1481866) to change this.

**Comment by emilio** [comment]
Blink also implemented (https://bugs.chromium.org/p/chromium/issues/detail?id=833105 is fixed), so I filed https://bugs.chromium.org/p/chromium/issues/detail?id=872356 for them.

**Comment by bradkemper** [comment]
Currently, the 2-axis starts-with-horizontal rule, and the up-to-four-compass-directions starts-with-top rule is understandable and mostly internally consistent. Having some values that follow those rules and others that don’t is going to be horribly confusing and hard to remember. I should have said so in the call.

**Comment by tabatkins** [comment]
Right, but it'll be confusing *no matter what*, as we have several properties already doing 2-axis in the logical order.  The physical order is more or less just a legacy mistake at this point.

**Comment by frivoal** [comment]
I was about to patch the spec, but now I'm confused: is the shorthand supposed to be overflow-block then overflow-inline (logical longhands), or overflow-y then overflow-x (physical longhands)? The resolution sounds like the former (especially since @fantasai invokes consistency with scroll-snap-align as a rationale), but as far as I can tell the code change by Mozilla  and the request in Chrome's bug tracker to match the resolution are the later.

It seems to me that they're wrong, but I wasn't there during the call, so I may be missing something. Which is it?

**Comment by Loirooriol** [comment]
All shorthands that don't contain `block` or `inline` in their name seem to correspond to physical shorthands, even if they are new like `inset`. But probably the same can be said about the order (`scroll-snap-align` is not a shorthand).

**Comment by frivoal** [comment]
that seems to contradict what @tabatkins and @fantasai were saying thought.

> for any new 2-axis properties we're trying to map them to the logical longhands

> important to keep scroll-snap-align and overflow consistent

**Comment by tabatkins** [comment]
Yes, our entire argument, and the resolution, was to have the order be "block, then inline", not "x, then y".  (The resolution is... confusingly written, since it refers to "the overflow-x and overflow-y shorthand" rather than... the "overflow" property, which predates either.

> All shorthands that don't contain block or inline in their name seem to correspond to physical shorthands

Incorrect (see any of the Grid properties). And doesn't make a ton of sense, since a property that applies to both axises wouldn't specialize its name with "block" or "inline" - that's for sub-properties that are doing a single axis.

**Comment by frivoal** [comment]
Ok, so I'll fix the spec per the resolution (i.e. in terms of logical directions), and have pinged Chrome and Mozilla in their relevant bugs about the misunderstanding.

**Comment by emilio** [comment]
I think that's way riskier than what we interpreted. For example, code like this will stop working:

```js
  document.body.style.overflow = "hidden";
  document.body.style.overflowX // Will start returning "", but returned "hidden" up until now.
```

I'm not sure that's totally acceptable...

**Comment by emilio** [comment]
Also, we haven't implemented `overflow-inline` / `overflow-block` in Firefox yet, which means that what was stated as a tweak to an existing shorthand now requires implementing two properties and breaking the behavior of the pre-existing shorthand.

**Comment by frivoal** [comment]
Discussed this a bit on the call:

> fantasai: It's block then inline, not y then x
> emilio: We're changing the shorthands the property expands to?
> emilio: Like, the longhands. It's way more risky then a tweak to shorthand. If you access via CSSOM it may break.
> fantasai: I see problem. For a one value value of overflow shorthand it has to be assigned to phsyical longhand. For 2 value it has to be assigned to logical. That would not break anything other then the last 4 months of work
> emilio: Right, but that i s not how any other shorthand works
> fantasai: We're intending the syntax change to happen, but haven't decided on it. For margin/border/padding there is intent to have a switch for if assigning to physical or logical. background-position is aligned to do  this.
> fantasai: There are 2 types  of values for background position. One set will assign to physical and the  other to logical
> florian: Why does it make a difference to physical or logical in case where assigning to long hand?
> fantasai: https://drafts.csswg.org/css-backgrounds-4/#the-background-position
> emilio: B/c you can read longhand in CSSOM
> emilio: If you read element/style you don't resolve the properties there
> TabAtkins: You don't resolve any...oh, you do shorthands. nevermind
> florian: I suggest we hash that out on github.

**Comment by Loirooriol** [comment]
> Incorrect (see any of the Grid properties)

Sure, grid properties refer to columns or rows which are logical, but there aren't physical properties. In cases with both physical and logical longhands, the shorthand seems to refer to the physical ones.

Changing `overflow` to refer to logical longhands is not backwards compatible, and I would avoid hacks to make it half physical half logical. The extra complexity for CSSOM doesn't seem worth it.

**Comment by tabatkins** [comment]
> In cases with both physical and logical longhands, the shorthand seems to refer to the physical ones.

Because all of those are legacy shorthands dating back before we started adding logical longhands. In the time since we started doing that, we've defaulted to logical, and in fact have generally done logical-**only**.

> Changing overflow to refer to logical longhands is not backwards compatible, and I would avoid hacks to make it half physical half logical. The extra complexity for CSSOM doesn't seem worth it.

In all current impls `overflow` only takes a single value; except for the shorthand-expansion-in-specified-style issue that Emilio brought up (thanks!), it's unobservable which pair of longhands it canonically expands to. We should think more about that one exposed spot.

> For example we need to define how the shorthand serialization argument works in that case. If I did: [snip]

Note that this isn't a backwards-compat question, as this is assuming use of the new logical longhands, nor it is even overflow-specific; this is just a general question about how shorthands with both physical and logical longhands should serialize.

(I'm not sure if this question is answered in css-logical offhand, as I'm not an editor - @fantasai or @frivoal?)

**Comment by Loirooriol** [comment]
> Because all of those are legacy shorthands dating back before we started adding logical longhands.

`inset` is added in CSS Logical and corresponds to physical longhands.

> except for the shorthand-expansion-in-specified-style

Yes, that's what I don't think it's backwards compatible.

> this is just a general question about how shorthands with both physical and logical longhands should serialize

Shorthands only correspond to physical or logical longhands, not both.

For example, `element.style.margin` will be empty string if you set logical margin longhands but not the physical ones. Or if you first set the physical ones and then you override them with logical ones, then `element.style.margin` will serialize the physical ones even if they are not being used. And if you set physical ones but not logical ones, `element.style.marginBlock` will be empty string.

That's how it has been implemented in Firefox and Blink. (Well, right now Blink never serializes logical shorthands, I'm writing a patch right now).

**Comment by fantasai** [comment]
> inset is added in CSS Logical and corresponds to physical longhands.

Yes, because it needed to be consistent with the rest of the margin/padding/border set.
`overflow` needs to be consistent with `scroll-snap-align`, more than anything.

> Shorthands only correspond to physical or logical longhands, not both.

Shorthands correspond to both. We're still working out the exact mechanism, but they do correspond to both.

> For example, element.style.margin will be empty string if you set logical margin longhands but not the physical ones. Or if you first set the physical ones and then you override them with logical ones, then element.style.margin will serialize the physical ones even if they are not being used. And if you set physical ones but not logical ones, element.style.marginBlock will be empty string.

Once we have a syntax for representing a logical `margin` declaration, it should return a value. (It's currently an open issue.)

**Comment by emilio** [comment]
I'm reverting the swap Firefox made in https://bugzilla.mozilla.org/show_bug.cgi?id=1492567. There's no point in shipping `overflow-y`, then `overflow-x` if it's neither what the WG wants (which is way more subtle), nor what other engines are going to ship (https://bugs.chromium.org/p/chromium/issues/detail?id=872356), nor what we've shipped already from Firefox 61 (which is overflow-x, then overflow-y).

**Comment by fantasai** [comment]
@emilio I would understand if you reverted to only accepting one value, but reverting to the opposite of the spec says means that we will end up with Web content depending on what you are shipping. I don't think this is a good plan. Either ship something compatible with the spec or object to the spec being incompatible with what you insist on shipping.

**Comment by emilio** [comment]
I replied on the bug but:

> @emilio I would understand if you reverted to only accepting one value, but reverting to the opposite of the spec says means that we will end up with Web content depending on what you are shipping. I don't think this is a good plan. Either ship something compatible with the spec or object to the spec being incompatible with what you insist on shipping.

I'm reverting to what we (and other engines) are shipping, to avoid shipping something that is both incompatible with the spec _and_ with our previous behavior.

We've been shipping the shorthand in the x/y form since Firefox 61. If we want to unship that (which I could understand) I can do that, but I definitely would like that to go through the regular path and not getting the change in a beta without any nightly cycle. Also, we should ask the Blink folks to unship their shorthand too in that case.

My objection with what the CSSWG seemed to resolve (which both me and David misinterpreted), apart from the ones I've expressed in comments above, is mostly that the WG decided we wanted to change a shorthand to work like no other shorthand currently does (referring to some properties sometimes, and to others other times) after one implementation had already shipped the previously-resolved thing, which is the opposite behavior, and another had already implemented and cleared for shipping the same way.

**Comment by BillGoldstein** [comment]
> Guess it'd imply new overflow-inline & overflow-block longhands.
...which obviously already exist.

**Comment by fantasai** [comment]
@emilio If you have an objection to what's in the spec, you should raise that issue, not leave an offhand comment in a bug that you've implemented the opposite of the spec without any indication of what you expect the WG or other implementers to do about it. If you go back and read my comment, I said there were two conditions either of which are reasonable positions to take: a) You implement something compatible with the spec. b) You implement something else and ask the spec to change. Your comments read off as c) You implement something incompatible with the spec, and don't ask for the spec to change. I don't think c) is reasonable behavior and consider it bad form.

**Comment by majido** [comment]
@tabatkins  mentioned this bug today. We should try to change this in Blink before the usage grows large and the compat issue difficult to deal with. Note that we have [already switched](https://bugs.chromium.org/p/chromium/issues/detail?id=854379) `scroll-snap-align` to have the block/inline order so it makes sense to change overflow parsing to match.

At the moment I am trying to understand the current usage and the scope of potential breakage. At the moment I am looking at httparchive data ([bug](https://bugs.chromium.org/p/chromium/issues/detail?id=872356#c6)) since Blink [usage stats](https://www.chromestatus.com/metrics/css/popularity#overflow) are most likely not helpful since this overloads existing single-value `overflow`.  If you have ideas how to better understand the potential compat issue please chime in that bug.

**Comment by ewilligers** [comment]
@gregwhitworth Can https://developer.microsoft.com/en-us/microsoft-edge/platform/usage/css/overflow/ show how often a pair of values are used?

**Comment by jonjohnjohnson** [comment]
Any news here on the experiment/data? We still got x/y in the wild and just hoping for a shorthand similar to `inset` that can take the impending [logical keyword modifier](https://github.com/w3c/csswg-drafts/issues/1282).

**Comment by ewilligers** [comment]
Stats: https://chromestatus.com/metrics/feature/timeline/popularity/2900

**Comment by fantasai** [comment]
It seems that it's too late to change `overflow` away from "x y" ordering, regardless of whether it's a good idea.

I think the follow-up question is, can we / should we switch the order of `scroll-snap-align` to match `overflow` (at least in horizontal writing modes)? Or should it stay consistent with `place-*` (and inconsistent with `overflow`)?

Fwiw, the other 2-value properties that are following the "block inline" order are: `grid`/`grid-template`, `gap`, `place-content`/`place-self`/`place-items`.

**Comment by frivoal** [comment]
Unless we want to resolve that we're happy with the status quo, we probably could do with a use counter on scroll-snap-align, to see if it's even possible to change it.

**Comment by dbaron** [comment]
Do you need a use counter beyond https://chromestatus.com/metrics/css/popularity#scroll-snap-align ?  4.6% seems to suggest there's a pretty big chance it's not changeable.

**Comment by lilles** [comment]
Should this issue be closed with no change?
